src/DynamoCore/Engine/ShortestQualifiedNameReplacer.cs

```diff
@@ -108,84 +142,79 @@ private bool TryShortenClassName(IdentifierListNode node, out AssociativeNode sh
+            var symbolList =
+                matchingClasses.Select(matchingClass => new Symbol(matchingClass.Name));
+            var shortNames = Symbol.GetShortestUniqueNames(symbolList);
 
-            string shortName;
-            // if there is no class conflict simply use the class name as the shortest name
-            if (matchingClasses.Length == 1)
+            // remove hidden class if any from shortNames before proceeding
+            var hiddenClassNodes = matchingClasses.
+                Where(x => x.ClassAttributes?.HiddenInLibrary ?? false).ToList();
+            if(hiddenClassNodes.Any())
             {
-                shortName = className;
+                foreach (var hiddenClassNode in hiddenClassNodes)
+                {
+                    var keyToRemove = new Symbol(hiddenClassNode.Name);
+                    shortNames.Remove(keyToRemove);
+                }
             }
-            else
+
+            // Get the shortest name corresponding to the fully qualified name
+            if(shortNames.TryGetValue(new Symbol(qualifiedName), out shortName))
             {
-                shortName = resolver != null ? resolver.LookupShortName(qualifiedName) : null;
+                return CoreUtils.CreateNodeFromString(shortName);
+            }
 
-                if (string.IsNullOrEmpty(shortName))
+            // If shortName for fully qualified classname is not found, it could be a base class
+            // present in class hierarchy of any of the other matchingClasses, in which case
+            // set shortName to the one for the derived class.
+            var qualifiedClassNode = matchingClasses.FirstOrDefault(x => x.Name == qualifiedName);
+            var classHierarchies = matchingClasses.Where(x => x != qualifiedClassNode).
+                Select(y => classTable.GetClassHierarchy(y));
+            foreach (var hierarchy in classHierarchies)
+            {
+                // If A derives from B, which derives from C, the hierarchy for A 
+                // is listed in that order: [A, B, C], so we start searching in reverse order.
+                for (int i = hierarchy.Count - 1; i > 0; i--)
                 {
-                    // Use the namespace list as input to derive the list of shortest unique names
-                    var symbolList =
-                        matchingClasses.Select(matchingClass => new ProtoCore.Namespace.Symbol(matchingClass))
-                            .ToList();
-                    var shortNames = ProtoCore.Namespace.Symbol.GetShortestUniqueNames(symbolList);
-
-                    // Get the shortest name corresponding to the fully qualified name
-                    shortName = shortNames[new ProtoCore.Namespace.Symbol(qualifiedName)];
+                    if (hierarchy[i] == qualifiedClassNode)
+                    {
+                        shortName = shortNames[new Symbol(hierarchy[0].Name)];
```

Comment:
just double checking - this should always be `0`?

Reply:
Yes, we want the most derived class in that list.